linfa-clustering/src/gaussian_mixture/algorithm.rs

```diff
@@ -0,0 +1,494 @@
+            GmmInitMethod::Random => {
+                let mut resp = Array2::<f64>::random_using(
+                    (n_samples, hyperparameters.n_clusters()),
+                    Uniform::new(0., 1.),
+                    &mut rng,
+                );
+                let totals = &resp.sum_axis(Axis(1)).insert_axis(Axis(0));
+                resp = (resp.reversed_axes() / totals).reversed_axes();
+                resp.mapv(|v| F::from(v).unwrap())
+            }
+        };
+
+        let (mut weights, means, covariances) = Self::estimate_gaussian_parameters(
+            &observations,
+            &resp,
+            hyperparameters.covariance_type(),
+            hyperparameters.reg_covariance(),
+        );
+        weights = weights / F::from(n_samples).unwrap();
+
+        // GmmCovarType = full
+        let precisions = Self::compute_precision_cholesky_full(&covariances);
+
+        GaussianMixtureModel {
+            covar_type: *hyperparameters.covariance_type(),
+            weights,
+            means,
+            precisions,
+        }
+    }
+
+    fn estimate_gaussian_parameters<D: Data<Elem = F>>(
+        observations: &ArrayBase<D, Ix2>,
+        resp: &Array2<F>,
+        _covar_type: &GmmCovarType,
+        reg_covar: F,
+    ) -> (Array1<F>, Array2<F>, Array3<F>) {
+        let nk = resp.sum_axis(Axis(0)) + F::from(10.).unwrap() * F::epsilon();
+        let nk2 = nk.to_owned().insert_axis(Axis(1));
+        let means = resp.t().dot(observations) / nk2;
+        // GmmCovarType = Full
+        let covariances =
+            Self::estimate_gaussian_covariances_full(&observations, resp, &nk, &means, reg_covar);
+        (nk, means, covariances)
+    }
+
+    fn estimate_gaussian_covariances_full<D: Data<Elem = F>>(
+        observations: &ArrayBase<D, Ix2>,
+        resp: &Array2<F>,
+        nk: &Array1<F>,
+        means: &Array2<F>,
+        reg_covar: F,
+    ) -> Array3<F> {
+        let n_clusters = means.nrows();
+        let n_features = means.ncols();
+        let mut covariances = Array::zeros((n_clusters, n_features, n_features));
+        for k in 0..n_clusters {
+            let diff = observations - &means.slice(s![k..k + 1, ..]);
+            let m = diff.t().to_owned() * resp.slice(s![.., k]);
+            let mut cov_k = m.dot(&diff) / nk[k];
+            let diag = cov_k.diag().to_owned() + reg_covar;
+            cov_k.diag_mut().assign(&diag);
+            covariances.slice_mut(s![k, .., ..]).assign(&cov_k);
+        }
+        covariances
+    }
+
+    fn compute_precision_cholesky_full<D: Data<Elem = F>>(
+        covariances: &ArrayBase<D, Ix3>,
+    ) -> Array3<F> {
+        let n_clusters = covariances.shape()[0];
+        let n_features = covariances.shape()[1];
+        let mut precisions_chol = Array::zeros((n_clusters, n_features, n_features));
+        for (k, covariance) in covariances.outer_iter().enumerate() {
+            let cov: Array2<f64> = covariance.mapv(|v| v.into());
```

Comment:
does the cholesky runs into instabilities for a lower precision?

Reply:
Well... If I use directly <code>covariance</code> I run into the following compilation problem: 
```bash
$ cargo test test_centroids_prediction
   Compiling linfa-clustering v0.1.0 (D:\rlafage\workspace\linfa\linfa-clustering)

error[E0599]: no method named `cholesky` found for struct `ndarray::ArrayBase<ndarray::ViewRepr<&F>, ndarray::Dim<[usize; 2]>>` in the current scope
   --> linfa-clustering\src\gaussian_mixture\algorithm.rs:241:23
    |
241 |             match cov.cholesky(UPLO::Lower) {
    |                       ^^^^^^^^ method not found in `ndarray::ArrayBase<ndarray::ViewRepr<&F>, ndarray::Dim<[usize; 2]>>`

```
even with <code>use ndarray_linalg::cholesky::*</code> at the top.
My workaround was to introduce Into<f64>

Reply:
take a look at the definition [here](https://docs.rs/ndarray-linalg/0.12.1/ndarray_linalg/cholesky/trait.Cholesky.html), the `Float` type does not implement `ndarray_linalg` `F: Lapack+Scalar`, because that depends on the lapack backend. You can introduce an additional bound, i.e. `F: Float + Lapack + Scalar` to fix that

Reply:
Yes, thank you!